### PR TITLE
Delay threshold review PR

### PR DESCRIPTION
**Pull request by rahulKQL**
Added two thresholds in the `BatcherImpl` which will trigger automatic batching upon breach of any of these.
 - ElementCount: Number of elements queued up till now.
 - ElementBytes: Size of the accumulated elements.
 - The threshold can be set using `BatchingSettings`.

### CODE REVIEW

File: gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import com.google.common.base.Preconditions;
+import javax.annotation.Nullable;
+
+/**
+ * Represents the batching settings to use for an API method that is capable of batching.
+ *
+ * <p>Each batching client will have their own set of default values for these threshold, Which will
+ * define the safest behavior for their jobs.
+ *
+ * <p>The default instance of this settings are configured to <b>not</b> use batching (i.e. the
```
Comment (romlogic) [inline gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java]: these

File: gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+import com.google.api.core.BetaApi;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import javax.annotation.Nullable;
+
+/**
+ * Represents the batching settings to use for an API method that is capable of batching.
+ *
+ * <p>Each batching client will have their own set of default values for these threshold, Which will
```
Comment (romlogic) [inline gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java]: these thresholds, which will define...

File: gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+ * define the safest behavior for their jobs.
+ *
+ * <p>The default instance of this settings are configured to <b>not</b> use batching (i.e. the
+ * batch size threshold is 1). This is the safest default behavior, which has meaning in all
+ * possible scenarios. Users are expected to configure actual batching thresholds explicitly: the
+ * element count, the request bytes count.
+ *
+ * <p>Warning: With the wrong settings, it is possible to cause long periods of dead waiting time.
```
Comment (romlogic) [inline gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java]: With the incorrect settings, it is possible to...

File: gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java (hunk 1 above)
Comment (rahulKQL) [reply gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java]: I think we are talking about BatchingSettings class by "this".

**Comment by igorbernstein2** [comment]
re: race condition
I don't love the atomic loops. I think it makes it harder to reason about correctness. I think it would be a lot easier to understand if `BatcherImpl.add()` acquired a lock for instantiating the Batch and adding an element to it. And `BatcherImpl.sendBatch` acquired the same lock to swap in a new batch

re: scheduling flushes
I don't see the benefit in maintaining a scheduled future for the first element of each Batch. I think it will be much simpler to just schedule a periodic task on construction of the BatcherImpl to call `sendBatch` and to cancel it on close. I don't think we care about precise timing here

**Comment by rahulKQL** [comment]
@igorbernstein2  Thanks a lot, That for the clear direction. I have updated delay threshold with this approach although I will update new approaches on the original PR. 

Please ignore any changes/commits on this local PR.

**Pull request closed without merging**